src/platform_linux/shmem.c

```diff
@@ -0,0 +1,1152 @@
+   if ((size >= SHM_LARGE_FRAG_SIZE)
+       && ((retptr = platform_shm_find_free(
+               shminfo, size, objname, func, file, lineno))
+           != NULL))
+   {
+      return retptr;
+   }
+
+   _Static_assert(sizeof(void *) == sizeof(size_t),
+                  "check our casts are valid");
+   // Optimistically, allocate the requested 'size' bytes of memory.
+   retptr = __sync_fetch_and_add(&shminfo->shm_next, (void *)size);
+
+   if (shminfo->shm_next > shminfo->shm_end) {
+      // This memory request cannot fit in available space. Reset.
+      __sync_fetch_and_sub(&shminfo->shm_next, (void *)size);
+
+      platform_error_log(
+         "[%s:%d::%s()]: Insufficient memory in shared segment"
+         " to allocate %lu bytes for '%s'. Approx free space=%lu bytes."
+         " shm_num_frags_tracked=%u, shm_num_frags_inuse=%u (HWM=%u).\n",
+         file,
+         lineno,
+         func,
+         size,
+         objname,
+         shminfo->shm_free_bytes,
+         shminfo->shm_num_frags_tracked,
+         shminfo->shm_num_frags_inuse,
+         shminfo->shm_num_frags_inuse_HWM);
+      return NULL;
+   }
+   // Track approx memory usage metrics; mainly for troubleshooting
+   __sync_fetch_and_add(&shminfo->shm_used_bytes, size);
+   __sync_fetch_and_sub(&shminfo->shm_free_bytes, size);
+
+   if (size >= SHM_LARGE_FRAG_SIZE) {
+      platform_shm_track_alloc(shminfo, retptr, size);
+   }
+
+   // Trace shared memory allocation; then return memory ptr.
+   if (Trace_shmem || Trace_shmem_allocs
+       || (Trace_large_frags && (size >= SHM_LARGE_FRAG_SIZE)))
+   {
+      platform_shm_trace_allocs(shminfo,
+                                size,
+                                "Allocated new fragment",
+                                retptr,
+                                objname,
+                                func,
+                                file,
+                                lineno);
+   }
+   return retptr;
+}
+
+/*
+ * -----------------------------------------------------------------------------
+ * platform_shm_realloc() -- Re-allocate n-bytes from shared segment.
+ *
+ * Functionally is similar to 'realloc' system call. We allocate required # of
+ * bytes, copy over the old contents (if any), and do a fake free of the oldptr.
+ * -----------------------------------------------------------------------------
+ */
+void *
+platform_shm_realloc(platform_heap_id hid,
+                     void            *oldptr,
+                     const size_t     oldsize,
+                     const size_t     newsize,
+                     const char      *func,
+                     const char      *file,
+                     const int        lineno)
+{
+   void *retptr = splinter_shm_alloc(hid, newsize, "Unknown");
+   if (retptr) {
+
+      // Copy over old contents, if any, and free that memory piece
+      if (oldptr) {
+         memcpy(retptr, oldptr, oldsize);
+         splinter_shm_free(hid, oldptr, "Unknown");
```

Comment:
This `realloc()` logic ... is working in this part-1. I had to overhaul and correct this implementation in part-3, after the pull from `/main` for Rob's large-values support. For now, let this be as-is; it's a working version.